### DISCUSSION TITLE
binvar or sdpvar as an input variable of a function

### DISCUSSION THREAD

**Discussion by ellensonny** [yalmip/YALMIP#1330] [General discussion]
The problem occurs when I use binvar or sdpvar variables as input to the graph function. Some other functions that come with matlab, such as sum and max, can allow binvar or sdpvar as input parameters. What is the reason for this？ how to solve this kind of problem.

code:
clear;clc;
yalmip('clear');
rng default
mp=case118;
F=randi([0 2],1,118);

% 定义变量
branch=binvar(1,186);
bus=binvar(1,118);
beta=binvar(1,118);

point=1:118;
load_point=beta.*point;
load_point=load_point(load_point~=0);
start=mp.branch(:,1)'.*branch;
start=start(start~=0);
finish=mp.branch(:,2)'.*branch;
finish=finish(finish~=0);
G=graph(start,finish);

%%约束条件
cons=[cons conncomp(G)==1];
cons=[cons ismember(load_point,[start finish])];

**Comment by johanlofberg** [comment]
I don't see the graph command ever could work on something which involves sdpvars. max/min etc are function which can be modelled using LP representations etc which is what YALMIP derives when you send an sdpvar to it. graph is a functionality which does a lot of manipulations, checks, and stuff, and doesn't just compute something which can be represented directly. you can of course represent the same thing using sdpvars, but you cannot use the high-level command graph to do it (i.e. you most likely use the command sparse with suitable indicies as it sounds like you simply want to create some sparse parameterization)

**Reply by ellensonny** [reply]
Thanks for your reply.Since high-level command graph cannot work in this case,does it mean that I have to write a part of the code myself to replace the functions implemented by the graph function. Is this too complicated? Do you have any suggestions to solve this problem?thanks.

**Reply by johanlofberg** [reply]
But does the graph function actually compute anything? To me it feels like it is just syntactic sugar to hold a sparse matrix, i.e. something like

mypattern = randn(n) > 0.1
X = binvar(n,n,'full').*mypattern

or same thing using the function sparse (which just is more complicated)
[i,j] = find(mypattern)
X = sparse(i,j,binvar(length(i),1),n,n)

**Reply by ellensonny** [reply]
Firstly,use the graph function to create a graph class, and then use the conncomp function to determine whether the graph is connected.
write in this way,it works.
code:
G=graph([1 2 3 5],[2 3 4 1]);
all(conncomp(G)==1);

 A sparse matrix could simplify calculations.But can it help me build simple optimization models? Implement the function of checking whether the graph is connected.My core problem is to test whether the graph is connected.
i try to replace[1 2 3 5],[2 3 4 1] with sdpvar or intvar binvar variables.as you see ,the  graph function cannot accept these variables as input arguments.

Besides,  i wonder whether yalimp could provide something similar to the fcn2optimexpr function to convert a function into an optimization expression.
 
Hope you can understand what I mean, thank you very much.

**Reply by johanlofberg** [reply]
That is way beyond what you use a modelling language for an optimization model for. You will have to formulate your model using standard MATLAB operators. To add a constraint that a graph is connected, you will have to express that as some kind of structured constraint (linear, sdp etc) on the elements on the graph

YALMIP version of fcn2optimexpr is probably the blackbox operator, but it gives you a crap model which is completely unstructured and you basically need luck for it to be solved. You don't want to develop optimization models that way. You should be able to understand how connectedness actually can be expressed, i.e along the lines of https://math.stackexchange.com/questions/4203408/ensure-a-graph-is-connected-in-a-linear-programming-problem#:~:text=%E2%88%91i%E2%88%88S%20%E2%88%91j%E2%88%89S%20xij%20%E2%89%A5%201.%20%E2%88%91%20i%20%E2%88%88,graph%20cannot%20be%20split%20into%20multiple%20connected%20components.
